External wallet files

ryanofsky opened the pull request:
This change consists of three commits:

* The first commit is a pure refactoring that removes the restriction that two wallets can only be opened at the same time if they are contained in the same directory.
* The second commit removes the restriction that `-wallet` filenames can only refer to files in the `-walletdir` directory.
* The third commit makes second commit a little safer by changing bitcoin to create wallet databases as directories rather than files, so they can be safely backed up.

All three commits should be straightforward:

*  The first commit adds around 20 lines of new code and then updates a bunch of function signatures (generally updating them to take plain fs::path parameters, instead of combinations of strings, fs::paths, and objects like CDBEnv and CWalletDBWrapper).
* The second commit removes two `-wallet` filename checks and adds some test cases to the multiwallet unit test.
* The third commit just changes the mapping from specified wallet paths to bdb environment & data paths.

  ---

**Note:** For anybody looking at this PR for the first time, I think you can skip the comments before _20 Nov_ and start reading at https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-345625565. Comments before  _20 Nov_ were about an earlier version of the PR that didn't include the third commit, and then confusion from not seeing the first commit.

ryanofsky committed:
Allow wallet files in multiple directories

Remove requirement that two wallet files can only be opened at the same time if
they are contained in the same directory.

This change mostly consists of updates to function signatures (updating
functions to take fs::path arguments, instead of combinations of strings,
fs::path, and CDBEnv / CWalletDBWrapper arguments).

ryanofsky committed:
Allow wallet files not in -walletdir directory

Remove restriction that -wallet filenames can only refer to files in the
-walletdir directory.

luke-jr commented:
This is too dangerous. It's not clear at face value to users that the `database/` directory will be created or must be maintained with the wallet file(s).

MeshCollider commented:
Agree with @luke-jr, #11446 is safer until we decide on a better wallet directory structure (e.g. https://github.com/bitcoin/bitcoin/pull/11466#issuecomment-335827526), which would involve supporting multiple BDB environments first.

MeshCollider commented:
Note that you might want to cherrypick https://github.com/bitcoin/bitcoin/pull/11466/commits/c36cb54711fe7677db6efc4db00a6e7d42e62f8d to help debug the travis failure

ryanofsky committed:
Create new wallet databases as directories rather than files

This change should make it easier for users to make complete backups of wallets
because they can now just back up the specified `-wallet=<path>` path directly,
instead of having to back up the specified path as well as the transaction log
directory (for incompletely flushed wallets).

Another advantage of this change is that if two wallets are located in the same
directory, they will now use their own BerkeleyDB environments instead using a
shared environment. Using a shared environment makes it difficult to manage and
back up wallets separately because transaction log files will contain a mix of
data from all wallets in the environment.

laanwj commented:
> FWIW, if you separate a wallet from its database directory you reliably get a wallet that won't open. 

Only if it has uncompacted log files. To prevent this, bitcoin spends so much time consolidating the wallet database at runtime (in the MaybeCompactWalletDB thread), and does it at least at a clean shutdown.

People are moving around wallet.dat already, from and to the data directory, so I'm not convinced that supporting multiple wallet directories in principle makes the risk much larger that wallets are separated from their database directory while "live". 

But indeed, a database environment has to be created where the wallets are. Creating wallets in different places or even on different file systems but having the database environment in the main data directory is asking for trouble.

promag commented on src/bitcoin-cli.cpp:
```diff
@@ -48,7 +48,7 @@ std::string HelpMessageCli()
     strUsage += HelpMessageOpt("-stdinrpcpass", strprintf(_("Read RPC password from standard input as a single line.  When combined with -stdin, the first line from standard input is used for the RPC password.")));
     strUsage += HelpMessageOpt("-stdin", _("Read extra arguments from standard input, one per line until EOF/Ctrl-D (recommended for sensitive information such as passphrases).  When combined with -stdinrpcpass, the first line from standard input is used for the RPC password."));
-    strUsage += HelpMessageOpt("-rpcwallet=<walletname>", _("Send RPC for non-default wallet on RPC server (argument is wallet filename in bitcoind directory, required if bitcoind/-Qt runs with multiple wallets)"));
+    strUsage += HelpMessageOpt("-rpcwallet=<walletname>", _("Send RPC for non-default wallet on RPC server (needs to exactly match corresponding -wallet option passed to bitcoind)"));
```
Full path? How will the url look like?

ryanofsky replied:
> Full path? How will the url look like?

It doesn't have to be the full path, but it can be. Paths are urlencoded, and symlinks are also allowed. See the unit test which tests every imaginable type of path.

promag replied:
Feels weird to known server side details. Not to mention it can break the client if you move the wallet to a new place.

ryanofsky replied:
> Feels weird to known server side details. Not to mention it can break the client if you move the wallet to a new place.

The PR is not changing anything about this. Whatever string the server uses to identify the wallet, the client just has to use the same string. We've discussed adding and storing wallet identifiers in the past, but opted to just use filenames for simplicity. Server operator is free to choose a path, or alias or whatever he wants as the string used to identify the wallet.

ryanofsky replied:
See discussion in #10650 "The walletID is currently defined by the filename (-wallet=<filename>), ideally, we later add a wallet RPC call to set the walletID (should be written to the wallet database)." It would still be possible to do this if you are interested, but it's tangential to this PR.

promag replied:
My point is that IMO we should not allow absolute path in rpcwallet, which you introduce here.

ryanofsky replied:
What is the problem exactly? If the server operator wants to use a plain path, or a relative path, he can use one. If he wants to use a symlink alias, he can use that. What is the problem with allowing him to use a full path as well?

Note that RPC identifier for wallet is entirely up to the server and not up the client. The client has to use whatever string the server uses to identify the wallet, there is no translation of any kind done on the client value.

promag replied:
I'm not arguing against full paths server side. I'm kind of against exposing absolute paths to RPC clients. `listwallets` will expose the full path right?

sipa replied:
It exposes the name the server uses for the wallet.

In the current implementation that name indeed corresponds to the configured path, but there is no reason why that must remain the case.

promag replied:
Well if the goal is to have some sort of id and have that exposed instead then +1.

ryanofsky replied:
> listwallets will expose the full path right?

No, or at least not by default. Listwallets just shows the wallet name. If the server is configured to use full paths for wallet name these will be full paths, but the server has to be specifically configured this way in order for this to happen.

I'm still little puzzled about what you think the underlying problem is with allowing server admins to use full paths as wallet names. But if you really think it is a problem it would literally be a one-line code change to use a different identifier. E.g. the `name` argument to `CreateWalletFromFile()` could be changed from `walletFile` to `fs::path(walletFile).filename()`.

promag replied:
Suppose you have configured the server with an absolute wallet path. Then you have a couple of clients (in different hosts). You had to configure each client with the absolute path of the wallet. These clients can be others than bitcoin-cli. Now the server operator decides it has to move the wallet to a new location. The clients can handle a server restart, but now the wallet doesn't exists. And the operator simply doesn't know of these clients because they are managed by other operators etc..

promag replied:
My point is to not even allow such cases. An wallet id (filename or even relative path) is enough

ryanofsky replied:
> Suppose you have configured the server with an absolute wallet path. Then you have a couple of clients (in different hosts). You had to configure each client with the absolute path of the wallet. These clients can be others than bitcoin-cli. Now the server operator decides it has to move the wallet to a new location. 

I don't see how this is much different than the server renaming the wallet. If a server admin wants to create a stable name for a wallet, it's perfectly possible to do that by creating a symlink or just manipulating the -datadir or -walletdir (#11687) settings to be able store the wallets in different locations without changing the name. I believe this PR actually gives server admins more flexibility in this regard rather than less, because previously wallet symlinks were disallowed.

But to take the example at face value, you have a server admin who accidentally exposed a non-stable wallet name, instead of a stable wallet name over the rpc interface to clients on the internet outside of his control. What solutions would you like to see for this problem? Would it be good enough to use `fs::path(walletFile).filename()` instead of `walletFile` as the wallet name passed to `CreateWalletFromFile` like I suggested earlier? Do you want wallet names to be configurable on the command line? Do you want them to be configurable by RPC? Do you want wallet names to be stored inside wallet databases like Jonas suggested?

promag replied:
> accidentally exposed a non-stable wallet

My point is to prevent this. IMO symlinks are a weak solution to remedy that case. Renaming is indeed a breaking change from the client point of view.

I'm not a server admin expert (not a bit either) but, correct me if I'm wrong, usually absolute paths aren't exposed outside.

ryanofsky commented:
> This is too dangerous. It's not clear at face value to users that the database/ directory will be created or must be maintained with the wallet file(s).

Note that this was posted in response to an earlier version of this pr that did not include the third commit. With the third commit, the default for creating new wallets is as directories rather than a files (with each wallet directory containing its own wallet.dat file, db.log and database/log files). See the comments and release notes in the third commit for details.

> But indeed, a database environment has to be created where the wallets are. Creating wallets in different places or even on different file systems but having the database environment in the main data directory is asking for trouble.

The PR doesn't do this (and never did). The first commit of the PR is a refactoring specifically to add support for opening multiple BDB environments.

> Agree with @luke-jr, #11466 is safer until we decide on a better wallet directory structure (https://github.com/bitcoin/bitcoin/pull/11466#issuecomment-335827526), which would involve supporting multiple BDB environments first.

Again, the very first commit of this PR adds support for multiple BDB environments, and if you combine this PR with #11466 the result is exactly what you described in your linked comment.

MeshCollider commented:
> Note that this was posted in response to an earlier version of this pr that did not include the third commit. With the third commit, the default for creating new wallets is as directories rather than a files (with each wallet directory containing its own wallet.dat file, db.log and database/log files). See the comments and release notes in the third commit for details.

Indeed, concept ACK on the new approach, will review soon

luke-jr commented:
IMO we should still forbid paths for file-based wallets, and only allow it for directory-based ones.

Suggest renaming `wallet.dat` in the directory to `wallet.bdb` to convey the fact that you can't just copy the file to make a proper backup.

sipa commented:
Concept ACK

laanwj commented:
> The PR doesn't do this (and never did). The first commit of the PR is a refactoring specifically to add support for opening multiple BDB environments.

Sorry! I didn't read the code but was apparently confused by all the screaming about dangerousness. If you add support for multiple database environments you get my concept ACK.

> Suggest renaming wallet.dat in the directory to wallet.bdb to convey the fact that you can't just copy the file to make a proper backup.

I don't think we should be changing the default naming of wallet databases. People have been backing up separate wallet.dats forever, and generally it works because bitcoin core leaves the wallet in consolidated state. So I'd prefer documenting that behavior instead.

Even a backup with the last records stripped is better than no backup. From that it's usually possible to recover due to the mempool or HD seed. But if they can't find 'wallet.dat', they might decide not to to a backup at all.

Turning the abstraction in the user's mind of wallets into directories is going to be an upstream fight as well. I'm not against it, but I think it will result in a lot of confusion.

ryanofsky commented:
> IMO we should still forbid paths for file-based wallets, and only allow it for directory-based ones.

Good idea, added check for this.

TheBlueMatt commented on src/wallet/db.h:
```diff
@@ -35,17 +35,18 @@ class CDBEnv
     void EnvShutdown();
 
 public:
-    mutable CCriticalSection cs_db;
```
Why make the lock global instead of per-env? It seems like there should be nothing preventing you from keeping that and adding a second lock that only exists to protect g_dbenvs, though likely not a big deal in terms of concurrency.

TheBlueMatt replied:
It just seems weird given it'd be even less code change to not split the lock out, and as long as the global lock is *only* held in the creation of new envs, no lockorder should be worried about (cause lockorder is always trivially correct if locks are held only for the duration of a function which uses them, without making callbacks).

TheBlueMatt commented on src/wallet/db.cpp:
```diff
@@ -81,29 +90,38 @@ void CDBEnv::Reset()
 
 void CDBEnv::Close()
 {
+    for (auto& db : mapDb) {
```
While you're at it, can you just merge EnvShutdown and Close? EnvShutdown is now only called from Close().

TheBlueMatt commented on src/wallet/db.cpp:
```diff
@@ -435,7 +467,9 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
                 throw std::runtime_error(strprintf("CDB: Error %d, can't open database %s", ret, strFilename));
             }
-            CheckUniqueFileid(*env, strFilename, *pdb_temp);
+            for (auto& env : g_dbenvs) {
```
Is this required? Somehow I had assumed they only conficted if they were both in the same env, but I dunno where I got that from.

TheBlueMatt replied:
Did you forget to push? Noting this is the only way we handle hardlinks is useful, indeed.

ryanofsky replied:
> Did you forget to push?

Oops, yeah. It was on the pr/wfile.9 tag I linked to below. Should be updated now. Also I'll work on the "just leave this part of the diff" change you requested above.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -3848,16 +3848,17 @@ std::vector<std::string> CWallet::GetDestValues(const std::string& prefix) const
 
-        std::unique_ptr<CWalletDBWrapper> dbw(new CWalletDBWrapper(&bitdb, walletFile));
-        std::unique_ptr<CWallet> tempWallet = MakeUnique<CWallet>(std::move(dbw));
+        std::unique_ptr<CWallet> tempWallet = MakeUnique<CWallet>(name, path);
```
Heh, I have to admit I really prefer the old version from the new one. Creating a DB and then giving it to the wallet makes more sense conceptually to me than telling the wallet where to find the DB and then having the CWallet constructor create the DB (especially as the slow-moving goal is to abstract the DB out of the wallet and make it an interface).

TheBlueMatt replied:
Yea, I really like that, tbh. If nothing else is much more clearly separates wallet.cpp from walletdb, etc, keeping wallet.cpp/CWallet only filled with wallet logic, and not database-backend stuff.

ryanofsky replied:
> If nothing else is much more clearly separates wallet.cpp from walletdb, etc, keeping wallet.cpp/CWallet only filled with wallet logic

It really affects callers of wallet, not wallet itself. CWallet is barely affected and wallet.cpp doesn't change at all. Anyway, let me know if you would prefer to see this here or in another PR.

TheBlueMatt replied:
I mean given you're changing it here, I see no reason to change it to go in the opposite direction of where I think it should go - even if CWalletDBWrapper right now is just a BDB implementation, it still *has* a clean-ish interface, and having the CWallet cosntructor create one from a path instead of take one just seems like the wrong direction. Better to just leave this part of the diff out of the PR?

ryanofsky replied:
Done in 39fa1e9c01417bc9275d8f81bfef67bfc5ca308e

ryanofsky reviewed (commented):
Added 2 commits 2cf4dc83328aae3ad120ca7b0a0dc39848346400 -> ec01271ca284a0ee73e58776cb3c651953c5536c ([pr/wfile.7](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.7) -> [pr/wfile.8](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wfile.7...pr/wfile.8))
Squashed ec01271ca284a0ee73e58776cb3c651953c5536c -> c25ed1fcecd31a5f4243205cdc43b645e2a8cdb6 ([pr/wfile.8](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.8) -> [pr/wfile.9](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.9))

TheBlueMatt commented on src/wallet/db.h:
```diff
@@ -100,9 +102,15 @@ class CWalletDBWrapper
+    CWalletDBWrapper(const fs::path& wallet_path, bool mock = false) :
+        nUpdateCounter(0), nLastSeen(0), nLastFlushed(0), nLastWalletUpdate(0)
     {
+        GetWalletEnv(wallet_path, env, strFile);
```
It just seems very weird given mock seems like it shouldn't have a/care what the wallet_path is. In the future I kinda feel like it should become more RAII (ie Open in the GetWalletEnv call) but not in this PR.

ryanofsky replied:
> It just seems very weird given mock seems like it shouldn't have a/care what the wallet_path is. In the future I kinda feel like it should become more RAII (ie Open in the GetWalletEnv call) but not in this PR.

RAII is not the model this uses because when there more than one wallet file open in the same directory they share the same environment. You could only resolve this by requiring all databases in the environment to be opened and closed at the same time (which would complicate dynamic loading/unloading) or by not allowing multiple databases in the same directory to be opened at the same time, or some combination these things, or just not using RAII.

On using fake "" path for mock databases, this is a little weird. Could be made less exposed by  adding a separate constructor for mocks like proposed in the other comment: https://github.com/bitcoin/bitcoin/pull/11687#discussion_r155371642

TheBlueMatt replied:
Yes, but instead of env = GetWalletEnv(wallet_path, strFile); env->Open(); or env->Close(); env->MakeMock();, GetWalletEnv could return a reference which is already open or otherwise has some level of understanding of closing when all refs to that env are lost. Even if not that, would prefer to not use the fake "" path as you suggest.

ryanofsky reviewed (commented):
Added 2 commits c25ed1fcecd31a5f4243205cdc43b645e2a8cdb6 -> b9440abd55804a6f001c58288cc7fd2124aa1007 ([pr/wfile.9](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.9) -> [pr/wfile.10](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wfile.9...pr/wfile.10))
Squashed b9440abd55804a6f001c58288cc7fd2124aa1007 -> 99934923d4c1dc4c0f96865b2589cc0b5df66b31 ([pr/wfile.10](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.10) -> [pr/wfile.11](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.11))

TheBlueMatt reviewed (commented):
utACK 99934923d4c1dc4c0f96865b2589cc0b5df66b31. Is a good improvement, but this stuff is quite sticky, so probably needs some additional pretty-through review.

ryanofsky commented on src/wallet/db.h:
```diff
@@ -100,9 +101,33 @@ class CWalletDBWrapper
     /** Create DB handle to real database */
-    CWalletDBWrapper(CDBEnv *env_in, const std::string &strFile_in) :
-        nUpdateCounter(0), nLastSeen(0), nLastFlushed(0), nLastWalletUpdate(0), env(env_in), strFile(strFile_in)
+    CWalletDBWrapper(const fs::path& wallet_path, bool mock = false) :
```
> Make the constructors private now?

I didn't do this just because it prevents being able to call MakeUnique so makes code below more verbose. But I think I'll wind up making this tweak if @promag or someone else asks for other style changes on the PR.

PierreRochard commented on src/wallet/db.cpp:
```diff
@@ -81,29 +101,24 @@ void CDBEnv::Reset()
     boost::this_thread::interruption_point();
 
-    strPath = pathIn.string();
+    fs::path pathIn = strPath;
```
Lines 32:33 in db.h states
`    // Don't change into fs::path, as that can result in
    // shutdown problems/crashes caused by a static initialized internal pointer.`
When originally written, this comment referred to boost::filesystem. I'm not sure if this comment is still relevant but I did want to bring attention to it.

ryanofsky replied:
> When originally written, this comment referred to boost::filesystem.

The comment is actually still referring to boost::filesystem, `fs` is just an alias: https://github.com/bitcoin/bitcoin/blob/62fdf9b07087b80d2142799bdd2324f61483359d/src/fs.h#L16

But it's fine to create a path object here. The comment is just explaining why the path objects aren't stored permanently (they stop being usable during shutdown).

PierreRochard commented on src/wallet/db.cpp:
```diff
@@ -53,20 +53,40 @@ void CheckUniqueFileid(const CDBEnv& env, const std::string& filename, Db& db)
+    fs::path env_directory = wallet_path.parent_path();
+    database_filename = wallet_path.filename().string();
+    LOCK(cs_db);
+    return &g_dbenvs.emplace(std::piecewise_construct, std::forward_as_tuple(env_directory.string()), std::forward_as_tuple(env_directory)).first->second;
```
style: tuple unpacking in the return statement reduces readability

ryanofsky replied:
> style: tuple unpacking in the return statement reduces readability

Unclear what preferred alternative is, but also this is replaced in aca8ec0dabb4987a4f73a19e2a7dc09177197adb from #11911 so it might not be relevant for long.

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/11687/commits/99934923d4c1dc4c0f96865b2589cc0b5df66b31

This is still on my to-do list for testing as well

jnewbery commented on src/wallet/init.cpp:
```diff
@@ -63,7 +63,7 @@ bool WalletParameterInteraction()
     }
 
-    gArgs.SoftSetArg("-wallet", DEFAULT_WALLET_DAT);
+    gArgs.SoftSetArg("-wallet", "");
```
This means that `m_name` is `""` if no `-wallet` name is specified on the command line or in the .conf file, which makes the `getwalletinfo` and `listwallets` output confusing:

```
bitcoin-cli getwalletinfo
{
  "walletname": "",
  "walletversion": 159900,
  "balance": 0.00000000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 0.00000000,
  "txcount": 0,
  "keypoololdest": 1513800503,
  "keypoolsize": 1000,
  "keypoolsize_hd_internal": 1000,
  "paytxfee": 0.00000000,
  "hdmasterkeyid": "4a4cbfcac33849b565a2eb5deecbc96c5c76c06a"
}

→ bitcoin-cli listwallets
[
  ""
]
```

Is it possible to change this so `m_name` is filled to `wallet.dat` when no `wallet` parameter is passed?

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -16,85 +16,101 @@ class MultiWalletTest(BitcoinTestFramework):
+        #   extern/w6  - to verify absolute wallet path is created correctly
+        #   w7_symlink - to verify symlinked wallet path is initialized correctly
+        #   w8         - to verify existing wallet file is loaded correctly
+        #   ''         - to verify default wallet file is created correctly
```
I don't think there's any way to access the 'default' wallet with name `''` through the CLI.

```
(Pdb) self.nodes[0].cli("--rpcwallet=").getwalletinfo()
*** subprocess.CalledProcessError: Command 'bitcoin-cli' returned non-zero exit status 18
```

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -16,85 +16,101 @@ class MultiWalletTest(BitcoinTestFramework):
+        for wallet_name, wallet in zip(wallet_names, wallets):
+            info = wallet.getwalletinfo()
+            assert_equal(info['immature_balance'], 50 if wallet is wallets[0] else 0)
+            assert_equal(info['walletname'], wallet_name)
```
Can you also add CLI testing to this test, to verify that the `-rpcwallet` argument still works? Something like:

```
self.nodes[0].cli("--rpcwallet=wallet_name").getwalletinfo()
```

jnewbery reviewed (commented):
Looks great. A few minor questions/suggestions inline.

sipa commented:
Code changes look good, and some of the refactors really help with readability.

Two things I'm unsure about:
* Making wallets directories. Maybe it results in slightly easier backup procedures (if people interpret it correctly), but it's also more mental load ("my wallet is just wallet.dat right?").
* Changing the wallet dir based on whether $DATADIR/wallets exists, that seems a bit fragile. Though it seems that's already in master?

MeshCollider commented:
@sipa yep that was part of #11466

ryanofsky reviewed (commented):
Rebased 99934923d4c1dc4c0f96865b2589cc0b5df66b31 -> 5c0d5c783ce97a2013219989bf1c6dd105916878 ([pr/wfile.11](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.11) -> [pr/wfile.12](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.12)) due to conflict in release notes.
Added 2 commits 5c0d5c783ce97a2013219989bf1c6dd105916878 -> 8e7981fe35f359fdbddd92f468443a61c844f392 ([pr/wfile.12](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.12) -> [pr/wfile.13](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wfile.12...pr/wfile.13))
Squashed 8e7981fe35f359fdbddd92f468443a61c844f392 -> fbc2402357dedd1e97cb2adcaef452393e656e8c ([pr/wfile.13](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.13) -> [pr/wfile.14](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.14))

ryanofsky commented:
> Two things I'm unsure about:
>
> - Making wallets directories. Maybe it results in slightly easier backup procedures (if people interpret it correctly), but it's also more mental load ("my wallet is just wallet.dat right?").

I'm not sure I understand the point about mental load.

For typical user who isn't using multiwallet or setting any `-wallet=path` option, nothing is changing here. wallet.dat is still created in the same place as before and can be backed up the same way. For more advanced users who do specify `-wallet=<path>` options, the specified path will just be created as a directory instead of a file.

> - Changing the wallet dir based on whether $DATADIR/wallets exists, that seems a bit fragile. Though it seems that's already in master?

Yes, this was done in #11466. I think a less fragile implementation of the idea might search for existing wallets in both `<walletdir>` and `<datadir>` but create all new wallets in `<walletdir>`. This could be implemented entirely in the wallet and would avoid the init order dependencies introduced in #11466 between node, qt, and wallet code. However, I don't actually understand what was accomplished in #11466 by changing the default wallet directory to begin with, so maybe this falls short in some way. Probably better to discuss somewhere else like #11466, though, since it's not really related to this PR.

promag commented on src/bitcoin-cli.cpp:
```diff
@@ -338,8 +338,8 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
-    std::string walletName = gArgs.GetArg("-rpcwallet", "");
-    if (!walletName.empty()) {
+    if (!gArgs.GetArgs("-rpcwallet").empty()) {
+        std::string walletName = gArgs.GetArg("-rpcwallet", "");
```
Why this change? To reduce `walletName` scope?

ryanofsky replied:
> Why this change? To reduce walletName scope?

Yes and to avoid declaring a variable that will never be used.

promag commented on test/functional/conf_args.py:
```diff
@@ -37,13 +37,13 @@ def run_test(self):
         self.start_node(0, ['-conf='+conf_file, '-wallet=w1'])
         self.stop_node(0)
-        assert os.path.isfile(os.path.join(new_data_dir, 'regtest', 'wallets', 'w1'))
+        assert os.path.exists(os.path.join(new_data_dir, 'regtest', 'wallets', 'w1'))
```
isdir?

ryanofsky replied:
Seems worse, makes test more fragile than it needs to be.

promag replied:
Yeah, agree, not relevant here.

promag reviewed (commented):
Testing.

ryanofsky commented:
Added 1 commit fbc2402357dedd1e97cb2adcaef452393e656e8c -> b355cf77508c5b2086062f151d83ff09b1371ca9 ([pr/wfile.14](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.14) -> [pr/wfile.15](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wfile.14...pr/wfile.15)) to fix travis test.
Squashed b355cf77508c5b2086062f151d83ff09b1371ca9 -> 303aadbd4de13b7956cf667d3bfad4310dc40564 ([pr/wfile.15](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.15) -> [pr/wfile.16](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.16))

sipa commented:
@ryanofsky Removed my concern about the wallets-becoming-directories. I guess that anyone who isn't aware that wallets can be directories will in fact not have any such wallets, as requires explicitly configuring a directory.

promag commented:
@ryanofsky needs rebase.

ryanofsky commented:
Rebased 5e8c79c1e8754bca52d7185983ba69dc9a33102f -> 2fd9c2b318b69ce3e7ce1fba5eb7f3fb5c781876 ([pr/wfile.16](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.16) -> [pr/wfile.17](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.17)) to avoid merge conflict in release notes. No changes.

ryanofsky commented:
Status of this PR is unclear, but more review is welcome. So far have:

Code ack MeshCollider Dec 18 https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-352574486
Code ack from TheBlueMatt Dec 13 https://github.com/bitcoin/bitcoin/pull/11687#pullrequestreview-83313824
Concept ack from sipa Nov 18 https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-345467887 and Dec 25 https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-353873641
Concept ack from laanwj Nov 20 https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-345625565
Partial reviews from jnewbery, promag, and PierreRochard.

There is a trivial merge conflict with segwit wallet #11403 in wallet_test_fixture.cpp, but this could probably be avoided if it matters.

jnewbery commented:
Sorry for the late concept feedback, but the config model seems a little confusing to me:

- `datadir` defaults to `~/.bitcoin` (on linux) and can be overridden by an absolute or relative path (relative to cwd)
- `walletdir` detaults to `<datadir>/wallets` if it exists, or `<datadir>` if it doesn't exist, and can be overridden by an absolute or relative path (relative to cwd)
- `wallet` defaults to `<walletdir>` and can be overridden by an absolute or relative path (relative to `<walletdir>`).

Without any arguments, a wallet will be created at `<datadir>/wallets/wallet.dat` (if `<datadir>/wallets` exists) or `<datadir>/wallet.dat` (if `<datadir>/wallets` does not exist). `listwallets` and `getwalletinfo` will show this wallet's name to be `''`. If bitcoind is restarted with `--wallet=wallet.dat`, `listwallets` and `getwalletinfo` will show the same wallet's name to be `wallet.dat`.

The main points of confusion:
1. `datadir` and `walletdir` paths are relative to cwd, but `wallet` is relative to `walletdir`
2. `walletdir` default depends on existence of `wallets` directory in `datadir`
3. wallet name can depend on how bitcoind was invoked

I think (1) and (2) could be addressed with very clear documentation (I think the release notes are *almost* there, but the help text could also be improved). You could also consider disallowing relative paths for `wallet` - it could either take a string (which would refer to a .dat file or directory in `walletdir`) or an absolute path.

I don't like (3) and think that the wallet name should always exactly equal the wallet file or directory.

jnewbery commented on src/wallet/db.h:
```diff
@@ -85,7 +85,8 @@ class CDBEnv
+CDBEnv* GetWalletEnv(const fs::path& wallet_path, std::string& database_filename);
 
 /** An instance of this class represents one database.
  * For BerkeleyDB this is just a (env, strFile) tuple.
```
https://github.com/bitcoin/bitcoin/pull/11687#discussion_r158147817

Yes, agree.

jnewbery commented on src/wallet/init.cpp:
```diff
@@ -33,7 +33,7 @@ std::string GetWalletHelpString(bool showDebug)
     strUsage += HelpMessageOpt("-walletrbf", strprintf(_("Send transactions with full-RBF opt-in enabled (RPC only, default: %u)"), DEFAULT_WALLET_RBF));
     strUsage += HelpMessageOpt("-upgradewallet", _("Upgrade wallet to latest format on startup"));
-    strUsage += HelpMessageOpt("-wallet=<file>", _("Specify wallet file (within data directory)") + " " + strprintf(_("(default: %s)"), DEFAULT_WALLET_DAT));
+    strUsage += HelpMessageOpt("-wallet=<path>", _("Specify wallet database path. Can be specified multiple times to load multiple wallets. Path is interpreted relative to the bitcoin data directory if it is not absolute, and will be created if it does not exist (as a directory containing a wallet.dat file and log files). For backwards compatibility this will also accept names of existing data fi ...
```
I believe this help text is incorrect. Should be `"Path is interpreted relative to the <walletdir> directory if it is not absolute"`

jnewbery commented on test/functional/multiwallet.py:
```diff
@@ -16,89 +16,105 @@ class MultiWalletTest(BitcoinTestFramework):
         self.stop_node(0)
 
-        self.start_node(0, self.extra_args[0])
+        self.start_node(0, extra_args)
```
nit: can combine with line above for `self.restart_node(0, extra_args)`

jnewbery reviewed (commented):
Tested ACK 2fd9c2b318b69ce3e7ce1fba5eb7f3fb5c781876.

This looks mostly good to me. A couple of comments inline. My only qualms are around the strange default wallet name `''` and perhaps slightly better documentation of the interactions between the various arguments.

ryanofsky reviewed (commented):
Thanks for the review and feedback!

Added 1 commits 2fd9c2b318b69ce3e7ce1fba5eb7f3fb5c781876 -> 69de7c1d8d8326d159634e8949bc3ac67266f534 ([pr/wfile.17](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.17) -> [pr/wfile.18](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wfile.17...pr/wfile.18))
Squashed 69de7c1d8d8326d159634e8949bc3ac67266f534 -> f9e7244449ad69b8b58bff4298bfacd9f16a1cd5 ([pr/wfile.18](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.18) -> [pr/wfile.19](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.19))

>Sorry for the late concept feedback, but the config model seems a little confusing to me:
>
>- `datadir` defaults to `~/.bitcoin` (on linux) and can be overridden by an absolute or relative path (relative to cwd)
>- `walletdir` detaults to `<datadir>/wallets` if it exists, or `<datadir>` if it doesn't exist, and can be overridden by an absolute or relative path (relative to cwd)
>- `wallet` defaults to `<walletdir>` and can be overridden by an absolute or relative path (relative to `<walletdir>`).
>
>Without any arguments, a wallet will be created at `<datadir>/wallets/wallet.dat` (if `<datadir>/wallets` exists) or `<datadir>/wallet.dat` (if `<datadir>/wallets` does not exist).

Everything described above was done in #11466. I definitely think some of the rough edges of that PR could be smoothed off, and have some ideas of how to simplify it, but this PR just takes the walletdir as is and doesn't change anything about how it is determined.

> `listwallets` and `getwalletinfo` will show this wallet's name to be `''`. If bitcoind is restarted with `--wallet=wallet.dat`, `listwallets` and `getwalletinfo` will show the same wallet's name to be `wallet.dat`.

Wallet naming is simple, I think. If you specify a `-wallet=FOO` option, the wallet name is `FOO`. If you don't specify a wallet option, the wallet name is the empty string. Whatever wallet name you specify is used verbatim everywhere.

>The main points of confusion:
>1. `datadir` and `walletdir` paths are relative to cwd, but `wallet` is relative to `walletdir`
>2. `walletdir` default depends on existence of `wallets` directory in `datadir`
>3. wallet name can depend on how bitcoind was invoked
>
>I think (1) and (2) could be addressed with very clear documentation (I think the release notes are *almost* there, but the help text could also be improved). You could also consider disallowing relative paths for `wallet` - it could either take a string (which would refer to a .dat file or directory in `walletdir`) or an absolute path.

I'm happy to write more documentation to address (1) and (2). I also think the behavior introduced in #11466 could be simplified, but that's seems more like a topic to discuss in #11466 or somewhere else.

> I don't like (3) and think that the wallet name should always exactly equal the wallet file or directory.

I think the wallet name should simply be whatever string the user specifies. That was the case before this PR and is still the case now. Other more complicated approaches are possible, and I wouldn't necessarily object to them, but it's not clear from your feedback here exactly what you have in mind.

ryanofsky commented:
Rebased f9e7244449ad69b8b58bff4298bfacd9f16a1cd5 -> 0a2d86f8ef64927c444dadcfc1f700705af784fa ([pr/wfile.19](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.19) -> [pr/wfile.20](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.20)) due to minor `wallet_test_fixture.cpp` conflict with #11403

ryanofsky commented:
Rebased 0a2d86f8ef64927c444dadcfc1f700705af784fa -> 5c8398dc03167f85ce5c8454fad8420dbcb89edb ([pr/wfile.20](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.20) -> [pr/wfile.21](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.21)) due to minor conflict with #12150

jnewbery commented:
As discussed, I think that the release notes in https://github.com/jnewbery/bitcoin/tree/pr11687.1 are perhaps clearer. I still think that the default `''` name is a little odd, but with those updated release notes, I think it makes sense.

ACK 5c8398dc03167f85ce5c8454fad8420dbcb89edb

jnewbery commented:
This now has three code ACKs and two concept ACKs. It'd be nice to get this into v0.16 if at all possible so we don't have two consecutive releases that change the wallet command line arguments.

ryanofsky commented:
> It'd be nice to get this into v0.16 if at all possible so we don't have two consecutive releases that change the wallet command line arguments.

Of course I'd also be happier to see this merged sooner, but I also don't think it's a big deal if this is pushed back. Unlike #11466, this doesn't change the location where files are created by default, so it's more backwards compatible than that change was. And despite some confusion in the comments above, I still see change and #11466 as basically orthogonal, since #11466 is more about where wallets are located, and this PR is more about how they are formatted.

alphaaurigae commented:
great, was exactly looking to have a wallet on a seperate luks drive.

ryanofsky commented:
> great, was exactly looking to have a wallet on a seperate luks drive.

This will be possible for you either with or without this PR in 0.16, because #11466, which is already merged, lets you specify a `-walletdir=<path>` option with the path to a directory or directory symlink. The directory can hold one or more bitcoin wallets loaded in a shared berkeleydb environment just like `-datadir` currently. #11466 also added the ability to create a `<datadir>/wallets` symlink instead of specifying a `-walletdir` option.

This PR extends #11466. It allows individual `-wallet=<name>` options to be paths, so if you have multiple wallets, they don't all have to live in the same directory and share a common berkeley db environment. It also creates new wallets (both inside and outside of `<walletdir>`) as directories instead of files, so they can be backed up independently and do not have mixed data logs.

ryanofsky commented:
Rebased 5c8398dc03167f85ce5c8454fad8420dbcb89edb -> 0797df3d2c06b132a911bed2e465834452f78d04 ([pr/wfile.21](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.21) -> [pr/wfile.22](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.22)) due to conflicts with #11970 in `multiwallet.py` test. No other changes.

jnewbery commented:
Tested ACK 0797df3d2c06b132a911bed2e465834452f78d04

Sjors commented:
It's quite useful to be able to keep the blockchain and wallet(s) in different locations. Using wallet directories also saves me trouble of remembering to type `.dat`.

Tested 0797df3d. Code changes look sane, but are a little over my head.

One surprising bit of behavior though is the difference between:

```sh
src/qt/bitcoin-qt -regtest -wallet=~/wallet
src/qt/bitcoin-qt -regtest -wallet=~/wallet.dat
```

The first command on OSX creates a `/home/Username/wallet` directory. The second creates a directory `~` inside the default regtest wallet directory.

Maybe we should disallow a file extension for new wallets?

ryanofsky commented:
> The first command on OSX creates a /home/Username/wallet directory. The second creates a directory ~ inside the default regtest wallet directory.

Interesting, what appears to be happening in that example is that your shell is expanding the tilde in the first command, but not expanding the tilde in your second command. (The PR doesn't implement any tilde expansion itself.) Testing locally with bash versions 4.4.12 and 4.3.48, I don't see tildes expanded in either of these cases. Bash documentation says that it only expands tildes at the beginning of words.

In any case, apart from the inconsistent tilde expansion, the behaviour you are describing doesn't actually seem surprising to me. If you specify a new path, and bitcoin goes and creates a directory at that path, this seems like pretty much the least surprising thing you could expect. But perhaps instead of allowing both absolute paths and paths relative to `<walletdir>` generally, the implementation could be restricted to only allow either absolute paths or paths **directly** in `<walletdir>`. Other options would be to forbid special characters or dots in paths like you suggested. I tend to be somewhat skeptical of approaches like these. It's annoying when software you are trying to use has random and unnecessary restrictions on what inputs it accepts. But if there's a realistic use-case where current behaviour would be harmful or confusing, adding some restrictions could make sense.

ryanofsky commented:
Rebased 0797df3d2c06b132a911bed2e465834452f78d04 -> 8a79f09a3dbcf48cfd8288ccf1d6bc3238aa1420 ([pr/wfile.22](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.22) -> [pr/wfile.23](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.23)) after #11904 merge. Resolved minor conflicts in CDBEnv::Open, no other changes.

Sjors commented:
Can you update the PR description reference  from `-datadir` to `-walletdir`?

>  the implementation could be restricted to only allow either absolute paths or paths directly in <walletdir>

Inclined to agree. See also discussion in #12166 about relative paths.

As you said, disallowing special characters doesn't seem like a good idea. Except perhaps disallowing `.dat` specifically for new wallets, or at least throwing a warning.

ryanofsky commented:
Rebased 8a79f09a3dbcf48cfd8288ccf1d6bc3238aa1420 -> 2f87a02746b6afee5a21e647c5b3159172736bbb ([pr/wfile.23](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.23) -> [pr/wfile.24](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.24)) due to conflict in release notes with #12210.

> Can you update the PR description reference from -datadir to -walletdir?

Good catch! Done.

>> the implementation could be restricted to only allow either absolute paths or paths directly in \<walletdir>
>
> Inclined to agree. See also discussion in https://github.com/bitcoin/bitcoin/pull/12166#discussion_r161561520 about relative paths.

Ok, I implemented this in 8c014825a6dba47f5a262da05e26394ef95b60df, but I didn't add the change this PR because I don't really like the complexity it adds and am not sure it is solving a real problem. To be clear, in the thread you linked to I was arguing that paths relative to $PWD were bad because they are unstable. I don't think think paths relative to walletdir really have the same problem. I'm going to open a new PR to address the relative -walletdir issue today, and I think this should also address the inconsistency John was concerned about.

> As you said, disallowing special characters doesn't seem like a good idea. Except perhaps disallowing .dat specifically for new wallets, or at least throwing a warning.

I'm not sure what the warning would be. It's not really clear to me what harm this would be trying to prevent.

Sjors commented:
>  It's not really clear to me what harm this would be trying to prevent.

Accidentally creating new empty wallets.

ryanofsky commented:
> Accidentally creating new empty wallets.

How exactly? If you specify the name of an existing file, it will just open the file.

Sjors commented:
People will accidentally use `.dat` for wallets that are directories and forget `.dat` for wallets that are files. They'll probably also create directories with `.dat` extension.

For this specific issue I would look for `[path].dat` if `path` doesn't  end with `.dat` and look for `[path - .dat]` if it does. A warning could be "Created a new wallet [path], but also found [path].dat".

Even better, but that would be breaking change, is to throw an error for non-existing wallets and introduce a new RPC method `-createwallet`.

ryanofsky commented:
Maybe we can discuss this in IRC, but a new wallet being created if you are trying to open a .dat file and forget the extension is something that happens today, and would always happen previously, unrelated to this PR. I think the changes you are talking about could be good idea but definitely belong in their own separate PR. It also seems to me this PR makes these problems less likely to occur going forward, because it allows wallets to be referenced as directories instead of as individual data files in a shared  environment.

jnewbery commented:
retested ACK ec4e3820a8a2351ba26c88dcc02d99e29bfe5524. Changes are just the minimum required for rebases.

ryanofsky commented:
Rebased 2f87a02746b6afee5a21e647c5b3159172736bbb -> ec4e3820a8a2351ba26c88dcc02d99e29bfe5524 ([pr/wfile.24](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.24) -> [pr/wfile.25](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.25)) due to release notes conflict with #12166.
Rebased ec4e3820a8a2351ba26c88dcc02d99e29bfe5524 -> aff4b44f0da4c583b76d8e3db926caf3fa03d3c9 ([pr/wfile.25](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.25) -> [pr/wfile.26](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.26)) due to release notes conflict with 0.16 version bump.
Rebased d7daf096830d17c4c20952ac5b487b49d43b4d80 -> 1fe5528f4dea8b530022b9bf359426f2079848af ([pr/wfile.26](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.26) -> [pr/wfile.27](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.27)) due to conflict with #12331.

promag reviewed (commented):
Tested ACK 1fe5528. I have two cases I'd like to discuss:

 1. should `-wallet=../foo` or `-wallet=foo/..` be possible?
 2. running `bitcoind -regtest -wallet=. -wallet` now gives:
```
************************
EXCEPTION: St13runtime_error
CDB: Can't open database wallet.dat (duplicates fileid 31e920000400000107a307f45b0d000000000000 from wallet.dat)
bitcoin in AppInit()
```
And before:
```
Error: Error loading wallet .. -wallet filename must be a regular file.
```

ryanofsky commented:
Rebased 1fe5528f4dea8b530022b9bf359426f2079848af -> 230f696ac105f41757c7e7625053941d483df0be ([pr/wfile.27](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.27) -> [pr/wfile.28](https://github.com/ryanofsky/bitcoin/commits/pr/wfile.28)) due to conflict with #12225.

>  I have two cases I'd like to discuss:

Could point out what you would like these test cases to do? They seem strange.

> should -wallet=../foo or -wallet=foo/.. be possible?

I don't see either a use-case here, or a reason to worry about the relative directory syntax, but if there is a concern you can articulate, adding more restrictions on path strings would be possible. You may want to look at 8c014825a6dba47f5a262da05e26394ef95b60df and surrounding discussion above.

> bitcoind -regtest -wallet=. -wallet

This is a weird invocation, and I'm not sure what you would want to happen with it. But the reason it fails to start with the "duplicates fileid" message is that this tries to load the default wallet twice, and any time you try to load two copies of the same wallet, you will see this.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -18,6 +18,7 @@
 #include <wallet/crypter.h>
 #include <wallet/walletdb.h>
 #include <wallet/rpcwallet.h>
+#include <util.h>
```
> Nit, sort :trollface:

Thanks, done in https://github.com/bitcoin/bitcoin/commit/c8c87343fcc7d86e9bec100a609a521c6c1e2d34

promag commented:
I mean, either the path is absolute or it's a name (which will be relative to `walletdir`). So anything relative (e.g. `.`, `..`, `../foo`, `foo/..`, ...) should fail. Unless there is a use case for relative paths?

ryanofsky commented:
> I mean, either the path is absolute or it's a name (which will be relative to walletdir). So anything relative (e.g. ., .., ../foo, foo/.., ...) should fail. Unless there is a use case for relative paths?

The use case for relative paths is pretty straightforward. Maybe you want to put your watch only wallets in a different directory than your wallets with private keys. Maybe you want to have different directories for different customers, or use nested directory structure, or use symlinks to avoid hardcoding paths that may change.

But I'm not sure if you are concerned about relative paths, or if you are concerned about particular types of path syntax like `..` or `.`, or something else. (It might help to state the concern explicitly.) If the concern is actually allowing `..` or `.` path syntax, I guess it would be possible to add operating-system specific code to forbid it, although I don't see what the point would be. It seems like it should be fine to just accept path strings and let the operating system / boost path library interpret them normally.

laanwj commented:
utACK be8ab7d082228d09ca529d1a08730d7d5aacb0ed

jimpo reviewed (commented):
Oops, post merge review.

MarcoFalke commented on src/wallet/init.cpp:
```diff
@@ -230,40 +230,44 @@ bool VerifyWallets()
+              (path_type == fs::symlink_file && fs::is_directory(wallet_path)) ||
+              (path_type == fs::regular_file && fs::path(walletFile).filename() == walletFile))) {
+            return InitError(strprintf(
+                _("Invalid -wallet path '%s'. -wallet path should point to a directory where wallet.dat and "
```
Very technical error message. Should not be translated, imo.

NickyYangYang commented:
hello,
      The wallet.dat size is so large.  I know there is so much keypairs in  the file. But , what is the structure of the file?  Is the same structure with the class CCryptoKeyStore?
